### ISSUE TITLE
[SB3] Request for Testing: Test latest releases with regards to detaching/re-attaching tablet/clipboard

### ISSUE THREAD

**Issue by qzed** [linux-surface/linux-surface#623]
Hi all,

the latest kernel release (v5.14.15) changes the way devices are handled when the base is detached from the clipboard/tablet part on the Surface Book 3. Normally, this should not be noticeable and everything should continue to work as before. However, detaching and re-attaching has been a bit unstable before, so it would be great if you give this a test and make sure that everything still works as it should be.

Specifically, detach and re-attach the tablet/clipboard part a couple of times both in laptop/normal and in reverse mode. While doing that please keep an eye on the kernel log (e.g. via `sudo dmesg -w`) and see if you can spot any errors or warnings. Also make sure that after re-attaching the keyboard and touchpad still work and the base battery shows up and has plausible values (e.g. run `acpi -b` or `upower -i $(upower -e | grep BAT2) | grep percentage` or have a look at the battery meter of the distro if it shows both batteries individually such as Gnome's Power Settings menu).

CC @Unknown-Fllooo, @shieldwed, @NickyTope

**Comment by shieldwed** [comment]
Hi @qzed 

Thanks a lot for your improvements. Here are my observations for the most recent kernel:

`Linux daniel-nb 5.14.15-arch1-1-surface #1 SMP PREEMPT Sun, 31 Oct 2021 21:25:33 +0000 x86_64 GNU/Linux`

1. I verified #621 as well and the dumped events are the same ✔️ 
2. keyboard and mouse are still working after detaching and reattaching the tablet in laptop mode ✔️ 
3. keyboard and mouse are not working after detaching and reattaching the tablet in reverse mode (this is expected though?) ✔️ 
4. both batteries are shown after detaching and reattaching ✔️ 
5. however, USB-C charging is not working anymore after detaching and reattaching. The USB-C cable needs to be removed and plugged in again to start charging again ❌ 
6. There are some errors in dmesg: ❌ 
```
...
[45990.565945] usb 3-2.1.5: USB disconnect, device number 9
[45990.573084] usb 3-2.2: USB disconnect, device number 6
[45990.573126] usb 3-2.2.3: USB disconnect, device number 8
[45990.965612] usb 3-2: new high-speed USB device number 13 using xhci_hcd
[45991.110489] usb 3-2: New USB device found, idVendor=045e, idProduct=0c12, bcdDevice=74.85
[45991.110530] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[45991.110549] usb 3-2: Product: USB2.1 Hub
[45991.110563] usb 3-2: Manufacturer: GenesysLogic
[45991.112743] hub 3-2:1.0: USB hub found
[45991.113149] hub 3-2:1.0: 3 ports detected
[45991.442331] usb 3-2.1: new high-speed USB device number 14 using xhci_hcd
[45991.446884] usb 3-2: USB disconnect, device number 13

[45991.446931] usb 3-2-port1: cannot reset (err = -71)
[45991.447499] usb 3-2-port1: attempt power cycle

[45992.021101] pcieport 0000:00:1d.4: pciehp: Slot(12): Link Down
[45992.021132] pcieport 0000:00:1d.4: pciehp: Slot(12): Card not present
[45992.021395] NVRM: Attempting to remove minor device 0 with non-zero usage count!
[45992.175651] usb 4-2: USB disconnect, device number 2
[45992.175687] usb 4-2.2: USB disconnect, device number 3

[45992.205911] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x0e, tc: 0x0e, tid: 0x01, cid: 0x1d, iid: 0x00)
[46001.815577] power_supply BAT2: driver failed to report `present' property: -110
[46017.353817] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x0e, tc: 0x0e, tid: 0x01, cid: 0x1d, iid: 0x00)

[46017.732138] usb 3-2: new high-speed USB device number 19 using xhci_hcd
[46017.875979] usb 3-2: New USB device found, idVendor=045e, idProduct=0c12, bcdDevice=74.85
[46017.876025] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[46017.876045] usb 3-2: Product: USB2.1 Hub
[46017.876059] usb 3-2: Manufacturer: GenesysLogic
[46017.877639] hub 3-2:1.0: USB hub found
[46017.877957] hub 3-2:1.0: 3 ports detected
...
```

```
....
[46057.618825] usb 3-2.2: USB disconnect, device number 21
[46057.618847] usb 3-2.2.3: USB disconnect, device number 23
[46058.018656] usb 3-2: new high-speed USB device number 28 using xhci_hcd

[46063.255419] usb 3-2: unable to get BOS descriptor set

[46063.257546] usb 3-2: New USB device found, idVendor=045e, idProduct=0c12, bcdDevice=74.85
[46063.257554] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[46063.257557] usb 3-2: Product: USB2.1 Hub
[46063.257560] usb 3-2: Manufacturer: GenesysLogic
[46063.258624] hub 3-2:1.0: USB hub found
[46063.259024] hub 3-2:1.0: 3 ports detected
[46063.582031] usb 3-2.1: new high-speed USB device number 29 using xhci_hcd
```

```
[46121.459439] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x02, tc: 0x02, tid: 0x02, cid: 0x01, iid: 0x01)

[46223.041936] INFO: task kworker/u16:10:129241 blocked for more than 122 seconds.
[46223.041950]       Tainted: P     U     OE     5.14.15-arch1-1-surface #1
[46223.041954] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[46223.041956] task:kworker/u16:10  state:D stack:    0 pid:129241 ppid:     2 flags:0x00004000
[46223.041967] Workqueue: kacpi_hotplug acpi_hotplug_work_fn
[46223.041977] Call Trace:
[46223.041984]  __schedule+0x333/0x1530
[46223.041998]  schedule+0x59/0xc0
[46223.042004]  schedule_preempt_disabled+0x11/0x20
[46223.042010]  __mutex_lock.constprop.0+0x2db/0x4a0
[46223.042019]  acpiphp_hotplug_notify+0x6d/0x280
[46223.042025]  ? acpiphp_post_dock_fixup+0xf0/0xf0
[46223.042030]  acpi_device_hotplug+0x99/0x3e0
[46223.042036]  acpi_hotplug_work_fn+0x3d/0x50
[46223.042040]  process_one_work+0x1e0/0x3b0
[46223.042048]  worker_thread+0x50/0x3b0
[46223.042053]  ? process_one_work+0x3b0/0x3b0
[46223.042058]  kthread+0x12f/0x160
[46223.042062]  ? set_kthread_struct+0x40/0x40
[46223.042066]  ret_from_fork+0x1f/0x30
[46345.921688] INFO: task kworker/u16:10:129241 blocked for more than 245 seconds.
[46345.921700]       Tainted: P     U     OE     5.14.15-arch1-1-surface #1
[46345.921703] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[46345.921706] task:kworker/u16:10  state:D stack:    0 pid:129241 ppid:     2 flags:0x00004000
[46345.921715] Workqueue: kacpi_hotplug acpi_hotplug_work_fn
[46345.921734] Call Trace:
[46345.921740]  __schedule+0x333/0x1530
[46345.921753]  schedule+0x59/0xc0
[46345.921758]  schedule_preempt_disabled+0x11/0x20
[46345.921765]  __mutex_lock.constprop.0+0x2db/0x4a0
[46345.921773]  acpiphp_hotplug_notify+0x6d/0x280
[46345.921778]  ? acpiphp_post_dock_fixup+0xf0/0xf0
[46345.921782]  acpi_device_hotplug+0x99/0x3e0
[46345.921788]  acpi_hotplug_work_fn+0x3d/0x50
[46345.921792]  process_one_work+0x1e0/0x3b0
[46345.921799]  worker_thread+0x50/0x3b0
[46345.921804]  ? process_one_work+0x3b0/0x3b0
[46345.921808]  kthread+0x12f/0x160
[46345.921811]  ? set_kthread_struct+0x40/0x40
[46345.921815]  ret_from_fork+0x1f/0x30

[46394.095000] usb 3-2.1: USB disconnect, device number 39
[46394.095004] usb 3-2.1.1: USB disconnect, device number 44
```
```
[46406.434119] hid-generic 0003:1038:1290.0041: hiddev100,hidraw9: USB HID v1.11 Device [SteelSeries Arctis Pro Wireless] on usb-0000:00:14.0-2.1.1.2/input1

[46468.801388] INFO: task kworker/u16:10:129241 blocked for more than 368 seconds.
[46468.801400]       Tainted: P     U     OE     5.14.15-arch1-1-surface #1
[46468.801404] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[46468.801406] task:kworker/u16:10  state:D stack:    0 pid:129241 ppid:     2 flags:0x00004000
[46468.801415] Workqueue: kacpi_hotplug acpi_hotplug_work_fn
[46468.801424] Call Trace:
[46468.801431]  __schedule+0x333/0x1530
[46468.801443]  schedule+0x59/0xc0
[46468.801449]  schedule_preempt_disabled+0x11/0x20
[46468.801455]  __mutex_lock.constprop.0+0x2db/0x4a0
[46468.801463]  acpiphp_hotplug_notify+0x6d/0x280
[46468.801468]  ? acpiphp_post_dock_fixup+0xf0/0xf0
[46468.801472]  acpi_device_hotplug+0x99/0x3e0
[46468.801478]  acpi_hotplug_work_fn+0x3d/0x50
[46468.801482]  process_one_work+0x1e0/0x3b0
[46468.801490]  worker_thread+0x50/0x3b0
[46468.801494]  ? process_one_work+0x3b0/0x3b0
[46468.801498]  kthread+0x12f/0x160
[46468.801502]  ? set_kthread_struct+0x40/0x40
[46468.801505]  ret_from_fork+0x1f/0x30

[46591.681235] INFO: task kworker/u16:10:129241 blocked for more than 491 seconds.
[46591.681247]       Tainted: P     U     OE     5.14.15-arch1-1-surface #1
[46591.681251] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[46591.681254] task:kworker/u16:10  state:D stack:    0 pid:129241 ppid:     2 flags:0x00004000
[46591.681263] Workqueue: kacpi_hotplug acpi_hotplug_work_fn
[46591.681274] Call Trace:
[46591.681280]  __schedule+0x333/0x1530
[46591.681292]  schedule+0x59/0xc0
[46591.681298]  schedule_preempt_disabled+0x11/0x20
[46591.681304]  __mutex_lock.constprop.0+0x2db/0x4a0
[46591.681312]  acpiphp_hotplug_notify+0x6d/0x280
[46591.681318]  ? acpiphp_post_dock_fixup+0xf0/0xf0
[46591.681322]  acpi_device_hotplug+0x99/0x3e0
[46591.681328]  acpi_hotplug_work_fn+0x3d/0x50
[46591.681332]  process_one_work+0x1e0/0x3b0
[46591.681340]  worker_thread+0x50/0x3b0
[46591.681345]  ? process_one_work+0x3b0/0x3b0
[46591.681349]  kthread+0x12f/0x160
[46591.681353]  ? set_kthread_struct+0x40/0x40
[46591.681357]  ret_from_fork+0x1f/0x30

[46714.561392] INFO: task kworker/u16:10:129241 blocked for more than 614 seconds.
[46714.561407]       Tainted: P     U     OE     5.14.15-arch1-1-surface #1
[46714.561412] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[46714.561415] task:kworker/u16:10  state:D stack:    0 pid:129241 ppid:     2 flags:0x00004000
[46714.561428] Workqueue: kacpi_hotplug acpi_hotplug_work_fn
[46714.561437] Call Trace:
[46714.561442]  __schedule+0x333/0x1530
[46714.561454]  schedule+0x59/0xc0
[46714.561460]  schedule_preempt_disabled+0x11/0x20
[46714.561490]  __mutex_lock.constprop.0+0x2db/0x4a0
[46714.561508]  acpiphp_hotplug_notify+0x6d/0x280
[46714.561513]  ? acpiphp_post_dock_fixup+0xf0/0xf0
[46714.561517]  acpi_device_hotplug+0x99/0x3e0
[46714.561523]  acpi_hotplug_work_fn+0x3d/0x50
[46714.561527]  process_one_work+0x1e0/0x3b0
[46714.561534]  worker_thread+0x50/0x3b0
[46714.561538]  ? process_one_work+0x3b0/0x3b0
[46714.561542]  kthread+0x12f/0x160
[46714.561546]  ? set_kthread_struct+0x40/0x40
[46714.561550]  ret_from_fork+0x1f/0x30

[46753.944884] usb 3-2: USB disconnect, device number 38
```
```
[46755.251567] usb 3-2.2: new high-speed USB device number 55 using xhci_hcd
[46755.294280] usb 3-2: USB disconnect, device number 53
[46755.348610] usb 3-2.2: Device not responding to setup address.

[46755.354770] hub 3-2.1:1.0: activate --> -19

[46755.454965] usb 4-2: USB disconnect, device number 6
[46755.454998] usb 4-2.2: USB disconnect, device number 7
[46755.495241] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x0e, tc: 0x0e, tid: 0x01, cid: 0x1d, iid: 0x00)

[46755.554926] usb 3-2.2: Device not responding to setup address.
[46755.761427] usb 3-2.2: device not accepting address 55, error -71

[46755.762133] usb 3-2-port2: attempt power cycle
[46756.159280] usb 3-2.1: USB disconnect, device number 54

[46764.908162] power_supply BAT2: driver failed to report `present' property: -110

[46780.255991] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x0e, tc: 0x0e, tid: 0x01, cid: 0x1d, iid: 0x00)
[46780.624874] usb 4-2: new SuperSpeed Plus Gen 2x1 USB device number 8 using xhci_hcd
[46780.644404] usb 4-2: New USB device found, idVendor=045e, idProduct=0c13, bcdDevice=74.85
[46780.644409] usb 4-2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[46780.644410] usb 4-2: Product: USB3.2 Hub
[46780.644411] usb 4-2: Manufacturer: GenesysLogic
[46780.647472] hub 4-2:1.0: USB hub found
[46780.647747] hub 4-2:1.0: 2 ports detected
[46780.764745] usb 3-2: new high-speed USB device number 59 using xhci_hcd
[46780.858296] snd_hda_intel 0000:00:1f.3: refused to change power state from D3hot to D0
[46780.912709] usb 3-2: New USB device found, idVendor=045e, idProduct=0c12, bcdDevice=74.85
[46780.912718] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[46780.912721] usb 3-2: Product: USB2.1 Hub
[46780.912723] usb 3-2: Manufacturer: GenesysLogic
[46780.913919] hub 3-2:1.0: USB hub found
[46780.914227] hub 3-2:1.0: 3 ports detected

[46780.980052] snd_hda_intel 0000:00:1f.3: CORB reset timeout#2, CORBRP = 65535

[46780.985376] usb 4-2.2: new SuperSpeed Plus Gen 2x1 USB device number 9 using xhci_hcd
[46781.005982] usb 4-2.2: New USB device found, idVendor=045e, idProduct=0c15, bcdDevice=74.97
[46781.005990] usb 4-2.2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[46781.005993] usb 4-2.2: Product: USB3.2 Hub
[46781.005996] usb 4-2.2: Manufacturer: GenesysLogic
[46781.011036] hub 4-2.2:1.0: USB hub found
[46781.011732] hub 4-2.2:1.0: 3 ports detected
[46781.238170] usb 3-2.1: new high-speed USB device number 60 using xhci_hcd

[46781.258370] snd_hda_codec_hdmi hdaudioC1D2: Unable to sync register 0x2f8100. -5

[46781.352793] usb 3-2.1: New USB device found, idVendor=0451, idProduct=8442, bcdDevice= 1.00
[46781.352802] usb 3-2.1: New USB device strings: Mfr=0, Product=0, SerialNumber=1
[46781.352805] usb 3-2.1: SerialNumber: 81070081CB11
[46781.354078] hub 3-2.1:1.0: USB hub found
[46781.354163] hub 3-2.1:1.0: 6 ports detected
[46781.451498] usb 3-2.2: new high-speed USB device number 61 using xhci_hcd

[46781.498327] snd_hda_codec_realtek hdaudioC1D0: Unable to sync register 0x2b8000. -5
[46781.498369] snd_hda_codec_realtek hdaudioC1D0: Unable to sync register 0x2b8000. -5

[46781.571838] usb 3-2.2: New USB device found, idVendor=045e, idProduct=0c14, bcdDevice=74.97
[46781.571856] usb 3-2.2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[46781.571891] usb 3-2.2: Product: USB2.1 Hub
[46781.571892] usb 3-2.2: Manufacturer: GenesysLogic
[46781.573505] hub 3-2.2:1.0: USB hub found
[46781.573901] hub 3-2.2:1.0: 3 ports detected
[46781.641466] usb 3-2.1.2: new high-speed USB device number 62 using xhci_hcd
```

```
[46837.441835] INFO: task kworker/u16:10:129241 blocked for more than 737 seconds.
[46837.441847]       Tainted: P     U     OE     5.14.15-arch1-1-surface #1
[46837.441850] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[46837.441853] task:kworker/u16:10  state:D stack:    0 pid:129241 ppid:     2 flags:0x00004000
[46837.441863] Workqueue: kacpi_hotplug acpi_hotplug_work_fn
[46837.441872] Call Trace:
[46837.441878]  __schedule+0x333/0x1530
[46837.441891]  schedule+0x59/0xc0
[46837.441896]  schedule_preempt_disabled+0x11/0x20
[46837.441903]  __mutex_lock.constprop.0+0x2db/0x4a0
[46837.441910]  acpiphp_hotplug_notify+0x6d/0x280
[46837.441916]  ? acpiphp_post_dock_fixup+0xf0/0xf0
[46837.441920]  acpi_device_hotplug+0x99/0x3e0
[46837.441926]  acpi_hotplug_work_fn+0x3d/0x50
[46837.441930]  process_one_work+0x1e0/0x3b0
[46837.441937]  worker_thread+0x50/0x3b0
[46837.441941]  ? process_one_work+0x3b0/0x3b0
[46837.441946]  kthread+0x12f/0x160
[46837.441949]  ? set_kthread_struct+0x40/0x40
[46837.441953]  ret_from_fork+0x1f/0x30
[46887.510078] audit: type=1106 audit(1636060830.267:715): pid=159238 uid=1000 auid=1000 ses=4 subj==unconfined msg='op=PAM:session_close grantors=pam_limits,pam_unix,pam_permit acct="root" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/0 res=success'
[46887.510088] audit: type=1104 audit(1636060830.267:716): pid=159238 uid=1000 auid=1000 ses=4 subj==unconfined msg='op=PAM:setcred grantors=pam_faillock,pam_permit,pam_env,pam_faillock acct="root" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/0 res=success'

[46960.322089] INFO: task kworker/u16:10:129241 blocked for more than 860 seconds.
[46960.322101]       Tainted: P     U     OE     5.14.15-arch1-1-surface #1
[46960.322105] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[46960.322108] task:kworker/u16:10  state:D stack:    0 pid:129241 ppid:     2 flags:0x00004000
[46960.322117] Workqueue: kacpi_hotplug acpi_hotplug_work_fn
[46960.322127] Call Trace:
[46960.322134]  __schedule+0x333/0x1530
[46960.322146]  schedule+0x59/0xc0
[46960.322151]  schedule_preempt_disabled+0x11/0x20
[46960.322158]  __mutex_lock.constprop.0+0x2db/0x4a0
[46960.322166]  acpiphp_hotplug_notify+0x6d/0x280
[46960.322171]  ? acpiphp_post_dock_fixup+0xf0/0xf0
[46960.322176]  acpi_device_hotplug+0x99/0x3e0
[46960.322181]  acpi_hotplug_work_fn+0x3d/0x50
[46960.322185]  process_one_work+0x1e0/0x3b0
[46960.322193]  worker_thread+0x50/0x3b0
[46960.322197]  ? process_one_work+0x3b0/0x3b0
[46960.322202]  kthread+0x12f/0x160
[46960.322205]  ? set_kthread_struct+0x40/0x40
[46960.322209]  ret_from_fork+0x1f/0x30

[46981.298311] NVRM: GPU at PCI:0000:02:00: GPU-3f617b0f-46b4-2946-1761-f685f70b6f40
[46981.298314] NVRM: Xid (PCI:0000:02:00): 79, pid=160602, GPU has fallen off the bus.
[46981.298316] NVRM: GPU 0000:02:00.0: GPU has fallen off the bus.
[46981.298450] NVRM: GPU 0000:02:00.0: GPU serial number is \xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff.
[46981.298457] NVRM: A GPU crash dump has been created. If possible, please run
               NVRM: nvidia-bug-report.sh as root to collect this data before
               NVRM: the NVIDIA kernel module is unloaded.
```
```
[47286.972444] usb 3-2.1.1.2: new full-speed USB device number 82 using xhci_hcd

[47292.272068] usb 3-2.1.1.2: unable to read config index 0 descriptor/start: -32
[47292.272080] usb 3-2.1.1.2: chopping to 0 config(s)
[47292.272085] usb 3-2.1.1.2: can't read configurations, error -32

[47292.342503] usb 3-2.1.1.2: new full-speed USB device number 83 using xhci_hcd
[47292.434764] usb 3-2.1.1.2: New USB device found, idVendor=1038, idProduct=1290, bcdDevice= 6.00
[47292.434773] usb 3-2.1.1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[47292.434777] usb 3-2.1.1.2: Product: Arctis Pro Wireless
[47292.434780] usb 3-2.1.1.2: Manufacturer: SteelSeries
[47292.442305] hid-generic 0003:1038:1290.0058: hiddev99,hidraw8: USB HID v1.11 Device [SteelSeries Arctis Pro Wireless] on usb-0000:00:14.0-2.1.1.2/input0
[47292.443570] hid-generic 0003:1038:1290.0059: hiddev100,hidraw9: USB HID v1.11 Device [SteelSeries Arctis Pro Wireless] on usb-0000:00:14.0-2.1.1.2/input1

[47328.962602] INFO: task kworker/u16:10:129241 blocked for more than 1228 seconds.
[47328.962614]       Tainted: P     U     OE     5.14.15-arch1-1-surface #1
[47328.962617] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[47328.962620] task:kworker/u16:10  state:D stack:    0 pid:129241 ppid:     2 flags:0x00004000
[47328.962629] Workqueue: kacpi_hotplug acpi_hotplug_work_fn
[47328.962639] Call Trace:
[47328.962645]  __schedule+0x333/0x1530
[47328.962657]  schedule+0x59/0xc0
[47328.962662]  schedule_preempt_disabled+0x11/0x20
[47328.962668]  __mutex_lock.constprop.0+0x2db/0x4a0
[47328.962676]  acpiphp_hotplug_notify+0x6d/0x280
[47328.962681]  ? acpiphp_post_dock_fixup+0xf0/0xf0
[47328.962686]  acpi_device_hotplug+0x99/0x3e0
[47328.962692]  acpi_hotplug_work_fn+0x3d/0x50
[47328.962696]  process_one_work+0x1e0/0x3b0
[47328.962703]  worker_thread+0x50/0x3b0
[47328.962707]  ? process_one_work+0x3b0/0x3b0
[47328.962712]  kthread+0x12f/0x160
[47328.962715]  ? set_kthread_struct+0x40/0x40
[47328.962719]  ret_from_fork+0x1f/0x30
```

I hope this is helpful.

**Comment by qzed** [comment]
Thanks, can you also
1. `git clone https://github.com/linux-surface/surface-aggregator-module` and `cd surface-aggregator-module`
2. Run `sudo modprobe surface_aggregator_registry` (this loads a debugging interface)
3. Run `sudo ./scripts/ssam/events.py listen 0x02,0x0e` (this starts logging SAM events for battery and KIP subsystem)
4. Detach and re-attach again (wait > 10s after detaching) until you can spot the `power_supply BAT2: driver failed to report present' property: -110` from your log above (this comes after a ~10s timeout runs out after detaching).
5. Once you've spotted that post the events from that detach process?

From the `power_supply BAT2: driver failed to report 'present' property: -110` line it seems that something is trying to access the battery after it has been detached, leading to `ETIMEDOUT` (errno 110).

The ACPI hotplug errors are... intersting. Looks like a deadlock. Any chance you had the nvidia GPU running in some way? Although that shouldn't have any impact on the ACPI hotplug code I think... should definitely be unrelated to SAM but I'll try to look into that.

Regarding USB-C charging, did that work before (or does it work on the official Arch kernel)? Could be related to the battery not being removed properly (i.e. the timeout error) but I somehow doubt that. There are also some messages related to USB, so it might be a separate problem unrelated to SAM.

**Comment by qzed** [comment]
Okay regarding the deadlock I suspect it hangs at [`pci_lock_rescan_remove()`](https://elixir.bootlin.com/linux/v5.14.15/source/drivers/pci/hotplug/acpiphp_glue.c#L788). There's also some locking with `acpi_lock_hp_context()` involved, but that's only for reference management, so as far as I can tell exclusively trivial stuff that only spans a couple instructions, so shouldn't be able to deadlock. It's possible that it's some other lock, but that should be further in and I think we'd see that in the trace. The `hotplug_event()` function is used only once in `acpiphp_hotplug_notify()` so that's likely inlined and why we don't see that in the trace.

I can't really tell where the second part to that lockup is though. That code runs on a workqueue (as far as I can tell listening to ACPI events) so it should either be a re-entrance issue somewhere down the line in that workqueue or another task is stuck, but that should then also show up in the logs.

**Comment by shieldwed** [comment]
Sorry for making you wait this long. Here are my findings nevertheless:
* `sudo modprobe surface_aggregator_registry` didn't provide me with the `/dev/surface/aggregator` device needed by `events.py`, however taking the command `sudo modprobe surface_aggregator_cdev` from #621 did. Here are the results:
  `events.py`:
  ```
  Event { time=22:12:10.960944, tc=02, tid=01, cid=53, iid=01, data=[01, b6, da, 00, 00, 96, 91, 00, 00, ae, 00, b4, 1f, 00, 00, 00, 00, e4, 25, 00, 00] }
  Event { time=22:12:12.956331, tc=02, tid=01, cid=53, iid=01, data=[01, 04, b0, 00, 00, fe, 88, 00, 00, ae, 00, 1e, 1e, 00, 00, 00, 00, e4, 25, 00, 00] }
  Event { time=22:12:13.954314, tc=02, tid=01, cid=53, iid=01, data=[01, 74, b8, 00, 00, ac, 8a, 00, 00, ae, 00, 9a, 1e, 00, 00, 00, 00, e4, 25, 00, 00] }
  Event { time=22:12:14.952088, tc=02, tid=01, cid=53, iid=01, data=[01, aa, be, 00, 00, ec, 8b, 00, 00, ae, 00, 6a, 1e, 00, 00, 00, 00, e4, 25, 00, 00] }
  Event { time=22:12:15.984125, tc=0e, tid=01, cid=2c, iid=00, data=[00] }
  Event { time=22:12:15.991222, tc=0e, tid=01, cid=1d, iid=00, data=[01] }
  Event { time=22:12:16.039024, tc=02, tid=01, cid=17, iid=01, data=[] }
  Event { time=22:12:22.024414, tc=02, tid=01, cid=53, iid=01, data=[01, 22, c4, 00, 00, 04, 8d, 00, 00, ae, 00, 94, 1e, 00, 00, 00, 00, e4, 25, 00, 00] }
  Event { time=22:12:27.016111, tc=02, tid=01, cid=53, iid=01, data=[01, aa, c8, 00, 00, f4, 8d, 00, 00, ae, 00, 65, 1e, 00, 00, 00, 00, e4, 25, 00, 00] }
  Event { time=22:12:34.014553, tc=02, tid=01, cid=53, iid=01, data=[01, 28, cd, 00, 00, da, 8e, 00, 00, ae, 00, b3, 1e, 00, 00, 00, 00, e4, 25, 00, 00] }
  Event { time=22:13:23.995415, tc=02, tid=01, cid=53, iid=01, data=[01, 38, d1, 00, 00, 7a, 8f, 00, 00, ac, 00, 55, 1e, 00, 00, 00, 00, e4, 25, 00, 00] }
  Event { time=22:14:01.887018, tc=0e, tid=01, cid=2c, iid=00, data=[00] }
  Event { time=22:14:01.890736, tc=0e, tid=01, cid=1d, iid=00, data=[03] }
  Event { time=22:14:03.404079, tc=02, tid=01, cid=53, iid=01, data=[01, d6, ce, 00, 00, 34, 8f, 00, 00, ae, 00, 93, 1d, e6, cd, 00, 00, e4, 25, 00, 00] }
  Event { time=22:14:06.075609, tc=02, tid=01, cid=53, iid=01, data=[01, ec, d6, 00, 00, ce, 90, 00, 00, ae, 00, 0b, 1f, e6, cd, 00, 00, e4, 25, 00, 00] }
  Event { time=22:14:06.424100, tc=0e, tid=01, cid=2c, iid=00, data=[01] }
  Event { time=22:14:06.425701, tc=0e, tid=01, cid=1d, iid=00, data=[03] }
  Event { time=22:14:06.977967, tc=02, tid=02, cid=17, iid=01, data=[] }
  Event { time=22:14:06.984533, tc=02, tid=02, cid=16, iid=01, data=[] }
  Event { time=22:14:06.992266, tc=02, tid=02, cid=15, iid=01, data=[] }
  Event { time=22:14:06.992397, tc=02, tid=02, cid=01, iid=01, data=[1f, 00, 00, 00] }
  Event { time=22:14:06.992439, tc=02, tid=02, cid=53, iid=01, data=[01, e6, 71, 02, 00, 82, 72, 01, 00, 7b, 00, 5e, 2f, 00, 00, 00, 00, c8, 32, 00, 00] }
  Event { time=22:14:28.985317, tc=02, tid=02, cid=53, iid=01, data=[01, f4, 6d, 02, 00, 84, 70, 01, 00, 7b, 00, ae, 2e, 00, 00, 00, 00, c8, 32, 00, 00] }
  Event { time=22:14:57.014139, tc=02, tid=01, cid=53, iid=01, data=[01, 20, d5, 00, 00, 74, 90, 00, 00, ae, 00, 5b, 1f, d4, cb, 00, 00, e4, 25, 00, 00] }
  ```
  `dmesg`:
  ```
  [ 3729.419404] usb 3-2: USB disconnect, device number 2
  [ 3729.419439] usb 3-2.1: USB disconnect, device number 4
  [ 3729.419454] usb 3-2.1.1: USB disconnect, device number 10
  [ 3729.419468] usb 3-2.1.1.2: USB disconnect, device number 12
  [ 3729.428788] usb 3-2.1.1.3: USB disconnect, device number 11
  [ 3729.440538] pcieport 0000:00:1d.4: pciehp: Slot(12): Link Down
  [ 3729.440643] NVRM: Attempting to remove minor device 0 with non-zero usage count!
  [ 3729.564806] usb 3-2.1.2: USB disconnect, device number 7
  [ 3729.637240] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x0e, tc: 0x0e, tid: 0x01, cid: 0x1d, iid: 0x00)
  [ 3729.642540] usb 3-2.1.5: USB disconnect, device number 9
  [ 3729.645539] usb 4-2: USB disconnect, device number 2
  [ 3729.645574] usb 4-2.2: USB disconnect, device number 3
  [ 3729.654745] usb 3-2.2: USB disconnect, device number 6
  [ 3730.398746] NVRM: GPU at PCI:0000:02:00: GPU-3f617b0f-46b4-2946-1761-f685f70b6f40
  [ 3730.398799] NVRM: Xid (PCI:0000:02:00): 79, pid=7408, GPU has fallen off the bus.
  [ 3730.398844] NVRM: GPU 0000:02:00.0: GPU has fallen off the bus.
  [ 3730.399105] NVRM: GPU 0000:02:00.0: GPU serial number is \xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff.
  [ 3730.399222] NVRM: A GPU crash dump has been created. If possible, please run
                 NVRM: nvidia-bug-report.sh as root to collect this data before
                 NVRM: the NVIDIA kernel module is unloaded.
  [ 3739.172071] power_supply BAT2: driver failed to report `present' property: -110
  ...
  [ 3835.537966] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x0e, tc: 0x0e, tid: 0x01, cid: 0x1d, iid: 0x00)
  [ 3835.935698] usb 3-2: new high-speed USB device number 13 using xhci_hcd
  [ 3836.082686] usb 3-2: New USB device found, idVendor=045e, idProduct=0c12, bcdDevice=74.85
  [ 3836.082743] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
  [ 3836.082772] usb 3-2: Product: USB2.1 Hub
  [ 3836.082794] usb 3-2: Manufacturer: GenesysLogic
  [ 3836.085188] hub 3-2:1.0: USB hub found
  [ 3836.095541] hub 3-2:1.0: 3 ports detected
  [ 3836.202623] usb 4-2: new SuperSpeed Plus Gen 2x1 USB device number 4 using xhci_hcd
  [ 3836.222589] usb 4-2: New USB device found, idVendor=045e, idProduct=0c13, bcdDevice=74.85
  [ 3836.222649] usb 4-2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
  [ 3836.222679] usb 4-2: Product: USB3.2 Hub
  [ 3836.222701] usb 4-2: Manufacturer: GenesysLogic
  [ 3836.226720] hub 4-2:1.0: USB hub found
  [ 3836.227099] hub 4-2:1.0: 2 ports detected
  [ 3836.435651] usb 3-2.1: new high-speed USB device number 14 using xhci_hcd
  [ 3836.556392] usb 3-2.1: New USB device found, idVendor=0451, idProduct=8442, bcdDevice= 1.00
  [ 3836.556458] usb 3-2.1: New USB device strings: Mfr=0, Product=0, SerialNumber=1
  [ 3836.556492] usb 3-2.1: SerialNumber: 81070081CB11
  [ 3836.559146] hub 3-2.1:1.0: USB hub found
  [ 3836.559289] hub 3-2.1:1.0: 6 ports detected
  [ 3836.632494] usb 4-2.2: new SuperSpeed Plus Gen 2x1 USB device number 5 using xhci_hcd
  [ 3836.652638] usb 4-2.2: New USB device found, idVendor=045e, idProduct=0c15, bcdDevice=74.97
  [ 3836.652704] usb 4-2.2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
  [ 3836.652741] usb 4-2.2: Product: USB3.2 Hub
  [ 3836.652759] usb 4-2.2: Manufacturer: GenesysLogic
  [ 3836.679869] hub 4-2.2:1.0: USB hub found
  [ 3836.680582] hub 4-2.2:1.0: 3 ports detected
  [ 3836.752342] usb 3-2.2: new high-speed USB device number 15 using xhci_hcd
  [ 3836.876137] usb 3-2.2: New USB device found, idVendor=045e, idProduct=0c14, bcdDevice=74.97
  [ 3836.876201] usb 3-2.2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
  [ 3836.876232] usb 3-2.2: Product: USB2.1 Hub
  [ 3836.876258] usb 3-2.2: Manufacturer: GenesysLogic
  [ 3836.879111] hub 3-2.2:1.0: USB hub found
  [ 3836.879548] hub 3-2.2:1.0: 3 ports detected
  [ 3836.945669] usb 3-2.1.2: new high-speed USB device number 16 using xhci_hcd
  [ 3837.306929] usb 3-2.1.2: New USB device found, idVendor=046d, idProduct=0836, bcdDevice= 0.10
  [ 3837.306933] usb 3-2.1.2: New USB device strings: Mfr=0, Product=2, SerialNumber=1
  [ 3837.306934] usb 3-2.1.2: Product: B525 HD Webcam
  [ 3837.306935] usb 3-2.1.2: SerialNumber: 9D14A7D0
  [ 3837.622273] usb 3-2.1.2: Warning! Unlikely big volume range (=6144), cval->res is probably wrong.
  [ 3837.622284] usb 3-2.1.2: [5] FU [Mic Capture Volume] ch = 1, val = 1536/7680/1
  [ 3837.622734] usb 3-2.1.2: Found UVC 1.00 device B525 HD Webcam (046d:0836)
  [ 3837.662331] input: B525 HD Webcam as /devices/pci0000:00/0000:00:14.0/usb3/3-2/3-2.1/3-2.1.2/3-2.1.2:1.2/input/input47
  [ 3837.735550] usb 3-2.1.5: new high-speed USB device number 18 using xhci_hcd
  [ 3837.825863] usb 3-2.1.5: New USB device found, idVendor=0451, idProduct=82ff, bcdDevice= 1.00
  [ 3837.825876] usb 3-2.1.5: New USB device strings: Mfr=0, Product=0, SerialNumber=1
  [ 3837.825881] usb 3-2.1.5: SerialNumber: 81070081CB11
  [ 3837.827831] hid-generic 0003:0451:82FF.0012: hiddev97,hidraw2: USB HID v1.11 Device [HID 0451:82ff] on usb-0000:00:14.0-2.1.5/input0
  [ 3838.032307] usb 3-2.1.1: new full-speed USB device number 19 using xhci_hcd
  [ 3838.123578] usb 3-2.1.1: New USB device found, idVendor=0451, idProduct=2036, bcdDevice= 1.01
  [ 3838.123584] usb 3-2.1.1: New USB device strings: Mfr=0, Product=1, SerialNumber=0
  [ 3838.123586] usb 3-2.1.1: Product: General Purpose USB Hub
  [ 3838.127908] hub 3-2.1.1:1.0: USB hub found
  [ 3838.128044] hub 3-2.1.1:1.0: 3 ports detected
  [ 3838.409029] usb 3-2.1.1.3: new full-speed USB device number 20 using xhci_hcd
  [ 3838.542239] usb 3-2.1.1.3: New USB device found, idVendor=1038, idProduct=1294, bcdDevice= 1.12
  [ 3838.542249] usb 3-2.1.1.3: New USB device strings: Mfr=4, Product=5, SerialNumber=0
  [ 3838.542252] usb 3-2.1.1.3: Product: Arctis Pro Wireless
  [ 3838.542255] usb 3-2.1.1.3: Manufacturer: SteelSeries 
  [ 3838.920338] input: SteelSeries  Arctis Pro Wireless Consumer Control as /devices/pci0000:00/0000:00:14.0/usb3/3-2/3-2.1/3-2.1.1/3-2.1.1.3/3-2.1.1.3:1.5/0003:1038:1294.0013/input/input48
  [ 3838.975949] input: SteelSeries  Arctis Pro Wireless as /devices/pci0000:00/0000:00:14.0/usb3/3-2/3-2.1/3-2.1.1/3-2.1.1.3/3-2.1.1.3:1.5/0003:1038:1294.0013/input/input49
  [ 3838.976236] hid-generic 0003:1038:1294.0013: input,hiddev98,hidraw3: USB HID v1.11 Device [SteelSeries  Arctis Pro Wireless] on usb-0000:00:14.0-2.1.1.3/input5
  [ 3839.835607] usb 3-2.1.1.2: new full-speed USB device number 21 using xhci_hcd
  [ 3839.930002] usb 3-2.1.1.2: New USB device found, idVendor=1038, idProduct=1290, bcdDevice= 6.00
  [ 3839.930011] usb 3-2.1.1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
  [ 3839.930014] usb 3-2.1.1.2: Product: Arctis Pro Wireless
  [ 3839.930016] usb 3-2.1.1.2: Manufacturer: SteelSeries
  [ 3839.938077] hid-generic 0003:1038:1290.0014: hiddev99,hidraw4: USB HID v1.11 Device [SteelSeries Arctis Pro Wireless] on usb-0000:00:14.0-2.1.1.2/input0
  [ 3839.938869] hid-generic 0003:1038:1290.0015: hiddev100,hidraw5: USB HID v1.11 Device [SteelSeries Arctis Pro Wireless] on usb-0000:00:14.0-2.1.1.2/input1
  [ 3840.072994] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x0e, tc: 0x0e, tid: 0x01, cid: 0x1d, iid: 0x00)
  [ 3840.386408] hid-generic 0019:045E:0C0A.0016: hidraw6: <UNKNOWN> HID v1.11 Device [Microsoft Surface 045E:0C0A] on 01:15:02:06:00
  [ 3840.393748] input: Microsoft Surface 045E:09AE Keyboard as /devices/platform/MSHW0117:00/01:0e:01:00:00/01:15:02:01:00/0019:045E:09AE.0017/input/input50
  [ 3840.394144] hid-generic 0019:045E:09AE.0017: input,hidraw7: <UNKNOWN> HID v1.11 Keyboard [Microsoft Surface 045E:09AE] on 01:15:02:01:00
  [ 3840.403143] hid-generic 0019:045E:0C0A.0018: hidraw8: <UNKNOWN> HID v1.11 Device [Microsoft Surface 045E:0C0A] on 01:15:02:05:00
  [ 3840.417555] input: Microsoft Surface 045E:09AF Mouse as /devices/platform/MSHW0117:00/01:0e:01:00:00/01:15:02:03:00/0019:045E:09AF.0019/input/input51
  [ 3840.417632] input: Microsoft Surface 045E:09AF Touchpad as /devices/platform/MSHW0117:00/01:0e:01:00:00/01:15:02:03:00/0019:045E:09AF.0019/input/input52
  [ 3840.417688] input: Microsoft Surface 045E:09AF UNKNOWN as /devices/platform/MSHW0117:00/01:0e:01:00:00/01:15:02:03:00/0019:045E:09AF.0019/input/input53
  [ 3840.417723] input: Microsoft Surface 045E:09AF UNKNOWN as /devices/platform/MSHW0117:00/01:0e:01:00:00/01:15:02:03:00/0019:045E:09AF.0019/input/input54
  [ 3840.417769] hid-multitouch 0019:045E:09AF.0019: input,hidraw9: <UNKNOWN> HID v1.11 Mouse [Microsoft Surface 045E:09AF] on 01:15:02:03:00
  [ 3840.639524] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x02, tc: 0x02, tid: 0x02, cid: 0x01, iid: 0x01)
  [ 3912.824987] gnome-documents[9943]: segfault at 20 ip 00007f3c18456c8e sp 00007fffe29f6620 error 4 in libtracker-sparql-2.0.so.0.306.0[7f3c1844d000+1c000]
  [ 3912.825000] Code: 01 00 00 4c 89 ff ff 15 c8 1c 02 00 4c 89 f7 ff 15 d7 16 02 00 0f 1f 80 00 00 00 00 4c 8b 74 24 28 4d 85 f6 0f 85 a1 fd ff ff <49> 8b 44 24 20 c7 00 01 00 00 00 48 83 7c 24 20 00 74 6f 48 8d 3d
  [ 3912.825027] audit: type=1701 audit(1636924519.172:257): auid=1000 uid=1000 gid=1000 ses=4 subj==unconfined pid=9943 comm="gnome-documents" exe="/usr/bin/gjs-console" sig=11 res=1
  [ 3912.872323] audit: type=1334 audit(1636924519.218:258): prog-id=39 op=LOAD
  [ 3912.874559] audit: type=1334 audit(1636924519.218:259): prog-id=40 op=LOAD
  [ 3912.874566] audit: type=1334 audit(1636924519.222:260): prog-id=41 op=LOAD
  [ 3912.874569] audit: type=1130 audit(1636924519.222:261): pid=1 uid=0 auid=4294967295 ses=4294967295 subj==unconfined msg='unit=systemd-coredump@1-10055-0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
  [ 3913.373122] audit: type=1131 audit(1636924519.722:262): pid=1 uid=0 auid=4294967295 ses=4294967295 subj==unconfined msg='unit=systemd-coredump@1-10055-0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
  [ 3913.489167] audit: type=1334 audit(1636924519.838:263): prog-id=41 op=UNLOAD
  [ 3913.489173] audit: type=1334 audit(1636924519.838:264): prog-id=40 op=UNLOAD
  [ 3913.489174] audit: type=1334 audit(1636924519.838:265): prog-id=39 op=UNLOAD
  [ 3922.300434] audit: type=1400 audit(1636924528.648:266): apparmor="STATUS" operation="profile_load" profile="unconfined" name="docker-default" pid=10553 comm="apparmor_parser"
  [ 3922.300439] audit: type=1300 audit(1636924528.648:266): arch=c000003e syscall=1 success=yes exit=7945 a0=4 a1=5579958eafc0 a2=1f09 a3=0 items=0 ppid=10552 pid=10553 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="apparmor_parser" exe="/usr/bin/apparmor_parser" subj==unconfined key=(null)
  [ 3922.300441] audit: type=1327 audit(1636924528.648:266): proctitle=61707061726D6F725F706172736572002D4B72002F7661722F6C69622F646F636B65722F746D702F646F636B65722D64656661756C7431343132303238373730
  [ 3922.353604] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
  [ 3922.355105] Bridge firewalling registered
  [ 3922.455188] audit: type=1325 audit(1636924528.802:267): table=nat:5 family=2 entries=2 op=nft_register_chain pid=10580 subj==unconfined comm="iptables"
  [ 3922.455198] audit: type=1300 audit(1636924528.802:267): arch=c000003e syscall=46 success=yes exit=116 a0=3 a1=7ffe2d0d9f70 a2=0 a3=7ffe2d0d9f5c items=0 ppid=673 pid=10580 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="iptables" exe="/usr/bin/xtables-nft-multi" subj==unconfined key=(null)
  [ 3922.455200] audit: type=1327 audit(1636924528.802:267): proctitle=2F7573722F62696E2F69707461626C6573002D773130002D74006E6174002D4E00444F434B4552
  [ 3922.464410] audit: type=1325 audit(1636924528.812:268): table=filter:6 family=2 entries=2 op=nft_register_chain pid=10584 subj==unconfined comm="iptables"
  [ 3922.464419] audit: type=1300 audit(1636924528.812:268): arch=c000003e syscall=46 success=yes exit=124 a0=3 a1=7ffdae43f490 a2=0 a3=7ffdae43f47c items=0 ppid=673 pid=10584 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="iptables" exe="/usr/bin/xtables-nft-multi" subj==unconfined key=(null)
  [ 3922.464423] audit: type=1327 audit(1636924528.812:268): proctitle=2F7573722F62696E2F69707461626C6573002D773130002D740066696C746572002D4E00444F434B4552
  [ 3922.474595] audit: type=1325 audit(1636924528.822:269): table=filter:7 family=2 entries=1 op=nft_register_chain pid=10588 subj==unconfined comm="iptables"
  [ 3932.026140] INFO: task kworker/u16:8:7712 blocked for more than 122 seconds.
  [ 3932.026153]       Tainted: P     U     OE     5.14.16-arch1-2-surface #1
  [ 3932.026157] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
  [ 3932.026159] task:kworker/u16:8   state:D stack:    0 pid: 7712 ppid:     2 flags:0x00004000
  [ 3932.026169] Workqueue: kacpi_hotplug acpi_hotplug_work_fn
  [ 3932.026179] Call Trace:
  [ 3932.026185]  __schedule+0x333/0x1530
  [ 3932.026207]  schedule+0x59/0xc0
  [ 3932.026213]  schedule_preempt_disabled+0x11/0x20
  [ 3932.026219]  __mutex_lock.constprop.0+0x2db/0x4a0
  [ 3932.026227]  acpiphp_hotplug_notify+0x6d/0x280
  [ 3932.026232]  ? acpiphp_post_dock_fixup+0xf0/0xf0
  [ 3932.026237]  acpi_device_hotplug+0x99/0x3e0
  [ 3932.026242]  acpi_hotplug_work_fn+0x3d/0x50
  [ 3932.026246]  process_one_work+0x1e0/0x3b0
  [ 3932.026253]  worker_thread+0x50/0x3b0
  [ 3932.026258]  ? process_one_work+0x3b0/0x3b0
  [ 3932.026262]  kthread+0x12f/0x160
  [ 3932.026265]  ? set_kthread_struct+0x40/0x40
  [ 3932.026269]  ret_from_fork+0x1f/0x30
  [ 3933.017110] kauditd_printk_skb: 187 callbacks suppressed
  ```
Stand by regarding USB-C charging.

**Comment by shieldwed** [comment]
Okay, regarding USB-C charging, this behaviour is neither exclusive to the linux-surface kernel nor to the Linux OS itself. The same behaviour actually exists on Windows. So, I guess it could even be a limitation on the hardware. Thanks for your thoughts on that anyway 👍🏾.

**Comment by shieldwed** [comment]
> Any chance you had the nvidia GPU running in some way?
According to `nvidia-smi` no processes are accessing the GPU. (However, I'm rather uncertain if this is really saying what you are asking for).

**Comment by Unknown-Fllooo** [comment]
> Hi all,
> 
> the latest kernel release (v5.14.15) changes the way devices are handled when the base is detached from the clipboard/tablet part on the Surface Book 3. Normally, this should not be noticeable and everything should continue to work as before. However, detaching and re-attaching has been a bit unstable before, so it would be great if you give this a test and make sure that everything still works as it should be.
> 
> Specifically, detach and re-attach the tablet/clipboard part a couple of times both in laptop/normal and in reverse mode. While doing that please keep an eye on the kernel log (e.g. via `sudo dmesg -w`) and see if you can spot any errors or warnings. Also make sure that after re-attaching the keyboard and touchpad still work and the base battery shows up and has plausible values (e.g. run `acpi -b` or `upower -i $(upower -e | grep BAT2) | grep percentage` or have a look at the battery meter of the distro if it shows both batteries individually such as Gnome's Power Settings menu).
> 
> CC @Unknown-Fllooo, @shieldwed, @NickyTope

I'm sorry i can't test these features anymore. I sent my Surface Book 3 back today, and I'm now waiting for the release of the laptop studio, you can tag me there about one week after the release in Europe. Looking forward to helping you guys testing, thanks for your great work btw. :)

**Comment by qzed** [comment]
> Sorry for making you wait this long.

No worries, I'm quite busy myself at the moment.

> Here are my findings nevertheless:
> 
> * `sudo modprobe surface_aggregator_registry` didn't provide me with the `/dev/surface/aggregator` device needed by `events.py`, however taking the command `sudo modprobe surface_aggregator_cdev` from [[SB3] Request for Testing: KIP Subsystem on Surface Book 3 #621](https://github.com/linux-surface/linux-surface/issues/621) did.

Oh right yes you're completely right. Sorry, that should have been `surface_aggregator_cdev`. Updated the comment above now.

Thanks for the logs! It seems that the events arrive in the correct order
```
# Some battery DPTF battery events (before detaching, it seems)
Event { time=22:12:10.960944, tc=02, tid=01, cid=53, iid=01, data=[01, b6, da, 00, 00, 96, 91, 00, 00, ae, 00, b4, 1f, 00, 00, 00, 00, e4, 25, 00, 00] }
Event { time=22:12:12.956331, tc=02, tid=01, cid=53, iid=01, data=[01, 04, b0, 00, 00, fe, 88, 00, 00, ae, 00, 1e, 1e, 00, 00, 00, 00, e4, 25, 00, 00] }
Event { time=22:12:13.954314, tc=02, tid=01, cid=53, iid=01, data=[01, 74, b8, 00, 00, ac, 8a, 00, 00, ae, 00, 9a, 1e, 00, 00, 00, 00, e4, 25, 00, 00] }
Event { time=22:12:14.952088, tc=02, tid=01, cid=53, iid=01, data=[01, aa, be, 00, 00, ec, 8b, 00, 00, ae, 00, 6a, 1e, 00, 00, 00, 00, e4, 25, 00, 00] }

# Events notifying that the base has been detached
Event { time=22:12:15.984125, tc=0e, tid=01, cid=2c, iid=00, data=[00] }
Event { time=22:12:15.991222, tc=0e, tid=01, cid=1d, iid=00, data=[01] }

# PSU/Adapter change event
Event { time=22:12:16.039024, tc=02, tid=01, cid=17, iid=01, data=[] }
```
As far as I can tell, the adapter change triggers a PSU changed notification in kernel and something then accesses the batteries. That creates a race with the devices being marked as hot-removed (and eventually being removed), which happens asynchronously. I think the fix should be fairly straightforward, just marking the devices as hot-removed synchronously, which should then prevent access like that.

**Comment by qzed** [comment]
Sorry it took me a bit, new kernels are building now. I've made the "mark as hot-removed" part synchronous, so hopefully that should improve things. Once v5.14.20/v5.15.3 are available, can you check if the battery issue is gone (specifically the `power_supply BAT2: driver failed to report 'present' property: -110` line)?

**Comment by shieldwed** [comment]
I just installed all updates:
```
Linux daniel-nb 5.15.3-arch1-1-surface #1 SMP PREEMPT Sat, 20 Nov 2021 03:24:49 +0000 x86_64 GNU/Linux
```
However, the issue still persists:
<details><summary>dmesg</summary>
<p>


```
[  768.723746] power_supply BAT2: driver failed to report `present' property: -110
...
[  982.910447] INFO: task kworker/u16:9:366 blocked for more than 122 seconds.
[  982.910457]       Tainted: P     U  W  OE     5.15.3-arch1-1-surface #1
[  982.910460] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[  982.910462] task:kworker/u16:9   state:D stack:    0 pid:  366 ppid:     2 flags:0x00004000
[  982.910469] Workqueue: kacpi_hotplug acpi_hotplug_work_fn
[  982.910479] Call Trace:
[  982.910481]  <TASK>
[  982.910485]  __schedule+0x331/0x1540
[  982.910492]  ? crypto_skcipher_encrypt+0x4c/0x70
[  982.910499]  schedule+0x5d/0xd0
[  982.910502]  schedule_preempt_disabled+0x11/0x20
[  982.910505]  __mutex_lock.constprop.0+0x260/0x460
[  982.910510]  acpiphp_hotplug_notify+0x6d/0x290
[  982.910517]  ? acpiphp_post_dock_fixup+0xf0/0xf0
[  982.910521]  acpi_device_hotplug+0x99/0x3e0
[  982.910527]  acpi_hotplug_work_fn+0x3d/0x50
[  982.910531]  process_one_work+0x1e5/0x3c0
[  982.910537]  worker_thread+0x50/0x3c0
[  982.910540]  ? process_one_work+0x3c0/0x3c0
[  982.910543]  kthread+0x12f/0x160
[  982.910548]  ? set_kthread_struct+0x50/0x50
[  982.910562]  ret_from_fork+0x1f/0x30
[  982.910570]  </TASK>
```

</p>
</details>

**Comment by qzed** [comment]
Thanks! Okay, seems like I need to have another look at this. I don't have much time at the moment so I'm likely going to revert the change for now.

**Comment by Boffice** [comment]
Hello,

I am using 5.15.3-surface on Zorin 16, screen is not responsive after detaching my SB2, and even keyboard and touchpad are unresponsive after re attaching it back.

**Comment by qzed** [comment]
The changes have been reverted in the latest release. I'll try to have a look at this again once I have more time.

**Comment by ZhiyuanChen** [comment]
Is there anything I can do to help?

**Comment by qzed** [comment]
Not at the moment.
